trie-db/src/lookup.rs

```diff
@@ -29,25 +29,49 @@ pub struct Lookup<'a, L: TrieLayout, Q: Query<L::Hash>> {
 	pub query: Q,
 	/// Hash to start at
 	pub hash: TrieHash<L>,
+	/// Layout for the trie.
+	pub layout: L,
 }
 
 impl<'a, L, Q> Lookup<'a, L, Q>
 where
 	L: TrieLayout,
 	Q: Query<L::Hash>,
 {
+	fn decode(mut self, v: Value, prefix: Prefix, depth: u32) -> Result<Option<Q::Item>, TrieHash<L>, CError<L>> {
+		match v {
+			Value::NoValue => Ok(None),
+			Value::Value(value) => Ok(Some(self.query.decode(value))),
+			Value::HashedValue(_, Some(value)) =>	Ok(Some(self.query.decode(value.as_slice()))),
+			Value::HashedValue(hash, None) => {
+				let mut res = TrieHash::<L>::default();
+				res.as_mut().copy_from_slice(hash);
+				if let Some(value) = self.db.get(&res, prefix) {
+					self.query.record(&res, &value, depth);
+					Ok(Some(self.query.decode(value.as_slice())))
+				} else {
+					Err(Box::new(TrieError::IncompleteDatabase(res)))
+				}
+			},
+		}
+	}
+
 	/// Look up the given key. If the value is found, it will be passed to the given
 	/// function to decode or copy.
 	pub fn look_up(
 		mut self,
 		key: NibbleSlice,
 	) -> Result<Option<Q::Item>, TrieHash<L>, CError<L>> {
 		let mut partial = key;
-		let mut hash = self.hash;
 		let mut key_nibbles = 0;
 
+		let mut full_key = key.clone();
+		full_key.advance(key.len());
+		let full_key = full_key.left();
```

Comment:
I don't understand this code, isn't key exactly same as full_key ? why not just this:
```suggestion
		let mut full_key = key.clone();
```

Reply:
key will be pointing to a different position in the slice than full_key.
Actually we advance by key base len so fullkey is the a nibble slice with position at end (size 0), and call to `left` will return the whole initial key.
It is the same as getting whole data used by the nibble slice, just there is no api for it.
(Value node are using their own full key as prefix so (b"key_to_value", None)).

Reply:
ok, makes sense